src/Microsoft.Extensions.Http.Polly/PolicyHttpMessageHandler.cs

```diff
@@ -77,7 +82,38 @@ protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage reques
                 throw new ArgumentNullException(nameof(request));
             }
 
-            return _policy.ExecuteAsync((ct) => base.SendAsync(request, ct), cancellationToken);
+            // Guarantee the existance of a context for every policy execution, but only create a new one if needed. This
+            // allows later handlers to flow state if desired.
+            var context = request.GetPollyContext();
+            if (context == null)
+            {
+                context = new Context(Guid.NewGuid().ToString());
```

Comment:
Suggestions how to create the executioncontextid?

Reply:
@rynowak Just for curiosity, why is there always a `Context` creation for each request execution (with `Context` and `string` allocation), even for uses that don't use it? Why didn't you prefer something like
```
public static Context GetPollyContext(this HttpRequestMessage request)
{
   if (!request.Properties.TryGetValue(PollyContextKey, out var context))
   {
       context = new Context(Guid.NewGuid().ToString());
       request.Properties[PollyContextKey] = context;
   }
   return context;
}
```

Reply:
`GetPollyContext()` is doing the `TryGet()` here and `SetPollyContext()` is going the set when one needs to be made. In that example it's not really getting, because it sets if it's not already there. Your example would be more `EnsurePollyContext()`.

Reply:
>Suggestions how to create the executioncontextid?

If the user has attached no context to the request, just:

    context = new Context();

[`Context.ExecutionKey`](https://github.com/App-vNext/Polly/blob/754a4a920900e1ae56d4fdae7cbdfd02bcb8cbd1/src/Polly.Shared/Context.cs#L41-L45) is the key [set via constructor](https://github.com/App-vNext/Polly/blob/754a4a920900e1ae56d4fdae7cbdfd02bcb8cbd1/src/Polly.Shared/Context.cs#L21-L24).  This represents the call site usage of the policy.  Not-set is allowed; if the user has not chosen to tag Polly logging/telemetry by call site tags (by using `request.SetPollyContext(new Context("MyUsageTag"));`),  we should not set it to something arbitrary.

_Compare:_ [`Context.ExecutionGuid`](https://github.com/App-vNext/Polly/blob/754a4a920900e1ae56d4fdae7cbdfd02bcb8cbd1/src/Polly.Shared/Context.cs) is the separate, per-execution-unique correlation id. That is [automatically initialised by Polly](https://github.com/App-vNext/Polly/blob/754a4a920900e1ae56d4fdae7cbdfd02bcb8cbd1/src/Polly.Shared/Context.cs#L51-L58).

Reply:
@Yves57 Re:

>why is there always a Context creation for each request execution [...] even for uses that don't use it? 
>Why didn't you prefer [...] so the context is allocated only if the user ask for it?

Polly uses a `Context` any time executing through a policy, to provide context to telemetry and logging, so the allocation isn't extra; Polly would allocate it later internally if it wasn't done here.

By scoping the context to the `HttpRequestMessage`, we ensure that the same `Context` instance is used through the whole chain of `DelegatingHandler`s when the user has configured multiple `PolicyHttpMessageHandler` instances.  This might be common, for example using both a circuit-breaker and retry policy: `.AddPolicyHandler(/* some circuit-breaker */).AddPolicyHandler(/* some retry policy */)`.  

While the motivation was to make that `Context` consistent through a chain of `PolicyHttpMessageHandler`s, it will actually mean fewer allocations: if `Context` wasn't scoped to `HttpRequestMessage`, the execution through each policy would generate a fresh `Context`, meaning more allocations.

Good question; hope this explanation helps.

Reply:
@reisenberger Thank you for the precise answer! Looking in the Polly code, I have seen that the "default" context that would be allocated. So I understand now that the best is to allocate it here.

Reply:
@reisenberger - FYI it will be `new Context(null)` the no-arg version is internal.